Ackermann Steering Plugin

knoedler opened the pull request:
## Summary
This adds an Ackermann Steering Plugin based on the differential drive plugin.

The plugin accepts a linear velocity (x) and a rotational velocity (z) the same as the differential drive plugin and sets the vehicle drive and steering joints to achieve the requested inputs.  The plugin and integration tests are largely based on the differential drive plugin.

Items to consider:
SpeedLimiter.cc and SpeedLimiter.hh are identical copies from the diff_drive plugin
Odometry is not as accurate as a pure diff_drive robot.  I had to open the integration tests limit up from 1cm to 25cm when comparing odometry to ground truth in the integration tests.  I tried various ground friction coefficients and using the wheel
slip plugin, but I did not achieve 1cm accuracy for the odometry.
There is a lot of commonality between the differential and Ackermann plugins.  It might make sense to merge them at some point in the future when the Ackermann plugin has aged a bit and is stable (to minimize risk to the heavily used diff drive plugin).
I was not able to generate a coverage report.  

## Test it
The plugin can be tested using examples/worlds/ackermann_steering.sdf


## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Added example world and/or tutorial
- [x] Updated documentation (as needed)
- [NA] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [?] While waiting for a review on your PR, please help review

ahcorde commented on src/systems/ackermann_steering/AckermannSteering.cc:
```diff
@@ -0,0 +1,771 @@
+    topics.push_back(_sdf->Get<std::string>("topic"));
+  }
+  topics.push_back("/model/" + this->dataPtr->model.Name(_ecm) + "/cmd_vel");
+  auto topic = validTopic(topics);
```
You should check if the output of validTopic is empty. https://ignitionrobotics.org/api/transport/9.0/classignition_1_1transport_1_1TopicUtils.html#a42548e276fce157d9794eb8a05bf07b3

here and below

knoedler replied:
Thank you for looking at the code.  All requested changes have been made and checked in.  codecheck and tests have been re-run and pass.
https://github.com/ignitionrobotics/ign-gazebo/pull/618/commits/7d56600edb0dd7d9695ee207a4f98a177f8b414e#diff-0dcac5ee6e5c74605877efaf6a30d6c6b5b23ef9d867f238e7a4b5185f1c4de0R327-R331

ahcorde commented on src/systems/ackermann_steering/AckermannSteering.cc:
```diff
@@ -0,0 +1,771 @@
+      std::chrono::duration_cast<std::chrono::steady_clock::duration>(odomPer);
+  }
+
+
```
```suggestion
```

ahcorde commented on src/systems/ackermann_steering/AckermannSteering.cc:
```diff
@@ -0,0 +1,771 @@
+    childFrame->add_value(this->sdfChildFrameId);
+  }
+
+
```
```suggestion
```

ahcorde commented on src/systems/ackermann_steering/AckermannSteering.cc:
```diff
@@ -0,0 +1,771 @@
+  this->leftJointSpeed =
+      (linVel * (1.0 - (this->wheelSeparation * tan(phi)) /
+                 (2.0 * this->wheelBase))) / this->wheelRadius;
+
```
```suggestion
```

ahcorde commented on src/systems/ackermann_steering/AckermannSteering.cc:
```diff
@@ -0,0 +1,771 @@
+  {
+    return;
+  }
+
```
```suggestion
```

ahcorde commented on src/systems/ackermann_steering/SpeedLimiter.cc:
```diff
@@ -0,0 +1,200 @@
+class ignition::gazebo::systems::SpeedLimiterPrivate
+{
+  /// \brief Class constructor.
+  public: SpeedLimiterPrivate(bool _hasVelocityLimits,
```
please document each argument

knoedler replied:
https://github.com/ignitionrobotics/ign-gazebo/pull/618/commits/7d56600edb0dd7d9695ee207a4f98a177f8b414e#diff-30d8d7c4b4f1b1762680c8a1895cb99472a7ddd049953fd6fab043d9d09157acR52-R60

ahcorde commented on test/integration/ackermann_steering_system.cc:
```diff
@@ -0,0 +1,429 @@
+  // Start server
+  ServerConfig serverConfig;
+  serverConfig.SetSdfFile(std::string(PROJECT_SOURCE_PATH) +
+      "/test/worlds/ackermann_steering.sdf");
```
please use `common::joinPaths`.

knoedler replied:
I addressed the three you found plus two others.
https://github.com/ignitionrobotics/ign-gazebo/pull/618/commits/7d56600edb0dd7d9695ee207a4f98a177f8b414e#diff-0dcac5ee6e5c74605877efaf6a30d6c6b5b23ef9d867f238e7a4b5185f1c4de0R327-R331

ahcorde commented on test/integration/ackermann_steering_system.cc:
```diff
@@ -0,0 +1,429 @@
+  // Start server
+  ServerConfig serverConfig;
+  serverConfig.SetSdfFile(std::string(PROJECT_SOURCE_PATH) +
+      "/test/worlds/ackermann_steering_slow_odom.sdf");
```
please use `common::joinPaths` to allow interoperability between OS

ahcorde commented on test/integration/ackermann_steering_system.cc:
```diff
@@ -0,0 +1,429 @@
+  // Start server
+  ServerConfig serverConfig;
+  serverConfig.SetSdfFile(std::string(PROJECT_SOURCE_PATH) +
+      "/test/worlds/ackermann_steering_custom_frame_id.sdf");
```
please use `common::joinPaths`

ahcorde commented on src/systems/ackermann_steering/AckermannSteering.hh:
```diff
@@ -0,0 +1,133 @@
+  /// https://github.com/ignitionrobotics/ign-gazebo/tree/ign-gazebo4/src/systems/diff_drive
+  /// https://www.auto.tuwien.ac.at/bib/pdf_TR/TR0183.pdf
+  /// https://github.com/froohoo/ackermansteer/blob/master/ackermansteer/
+
+
+
```
```suggestion

```

ahcorde commented:
@osrf-jenkins retest this please

chapulina reviewed (commented):
This is looking good to me! I didn't check the math too closely though. I took a quick look at the implementation on [gazebo_ros_pkgs](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/50c2d1b329936b2c60cdd78aabbf616ff691e43c/gazebo_plugins/src/gazebo_ros_ackermann_drive.cpp#L513-L534) but the approach looks a bit different.

ahcorde reviewed (changes requested):
Remember to sign all your commits. DCO is failing

knoedler commented:
> Remember to sign all your commits. DCO is failing

That particular commit was from clicking on "commit suggestion".  I would have expected github to handle signoffs.  I have made a couple of attempts to resolve it, but my googling has not found a successful way to resolve it.

I believe I have addressed the issues from the two reviews.

chapulina commented:
>  I would have expected github to handle signoffs. 

Yeah unfortunately GitHub doesn't make it easy to sign commits from the web UI. We've been manually adding the signature for those cases.

One way of solving this would be by adding the signature locally while rebasing and then force-pushing. If this is too much trouble, we can manually accept the DCO check since all other commits are properly signed.

knoedler commented:
> 
> One way of solving this would be by adding the signature locally while rebasing and then force-pushing. If this is too much trouble, we can manually accept the DCO check since all other commits are properly signed.

If you could, please manually accept the DCO check.

Adding the signature locally/rebasing/force-pushing was one of the approaches I tried a couple of different ways, but it didn't work for me in this case.

chapulina reviewed (approved):
Thank you for the new plugin!
